### PR TITLE
Proposed requirements and expanded spec for facet registrations

### PR DESCRIPTION
**Pull request by Sheeri**
#### One-line summary:

proposed updates to registry.md

### Problem

Call for contributions to the proposed registry

### Solution

Changes:
**02c1fc611a6005df3063d98475eb489f1e0b2d50** - Added a requirement

**b8f6e5b4adc11e5b910d89a08fed7a2585f61ec7** - Added an entry for "core" to propose changes to the spec with a "known" starting point.

This entry:

---
root_doc_URL: "https://openlineage.io/spec/facets/"
produced_facets: [
  "ol:core:1-0-0/ColumnLineageDatasetFacet.json",
  "ol:core:1-0-1/ColumnLineageDatasetFacet.json",
  "ol:core:1-0-0/DataQualityAssertionsDatasetFacet.json"
]

---

indicates that the documentation for the produced facets are at:
https://openlineage.io/spec/facets/1-0-0/ColumnLineageDatasetFacet.json
https://openlineage.io/spec/facets/1-0-1/ColumnLineageDatasetFacet.json
https://openlineage.io/spec/facets/1-0-0/DataQualityAssertionsDatasetFacet.json

**b099579e093e33461d4ba39c8f3de25c3fc518ef** - Proposing a new formatting of producer/consumer objects instead of an array of strings.

I have modified the "core" example, to show the translation between old and new format.
Added an "egeria" example which is both a producer and consumer.

- Producer and consumer doc root URLs may differ, so they are set inside the producer/consumer object
- It is assumed that the documentation link only applies to the facets owned by the same entity
  - e.g. egeria owns NewCustomFacet so the docs are at the Egeria doc URL
  - egeria does not own ColumnLineageDatasetFacet so there are no docs at the Egeria doc URL (unless it's extended by egeria?)
- "sample_URL" is where the examples/tests can be found (recommended but not required)
- "owner" is added for clarification. egeria produces their own custom facet plus one facet from core in this example
- spec_versions array for compatibility
- use_cases to better create the documentation page
- since there are "consumer" and "producer" objects, "facets" replaces "produced_facets" and "consumed_facets"

**8a0aee1b88d2aa326ef83022819d5b4a09bd60db** - added another proposed requirement 

**11da7b5a697ef4f0add1a61f4e6229418d9503ba** - added note about accuracy of registry entries

**a53cfa04c7d83c3517ff42dfa325e5ebb52a359b** - added Acceptance guidelines

**d9eacf244b98fd9b8df03d29edaa80757e2cc8b3** - added a note about reserving names for the future.

### Checklist
~~- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model~~
~~- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)~~

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
~~- [x] Your changes are accompanied by tests (_if relevant_)~~
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained - NOTE: each commit is self-contained, but the PR contains multiple commits. 
~~- [x] You've updated any relevant documentation (_if relevant_)~~
- [x] Your comment includes a one-liner for the changelog about the specific purpose of the change (_if necessary_)
~~- [x] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)~~
~~- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)~~

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: proposals/2161/registry.md [hunk 1]
```diff
@@ -18,6 +18,8 @@ Requirements:
 
 - Producers can create and evolve their custom facets without requiring approval from the OpenLineage project.
 - Producers and Consumers can update the list of the facets they produce or consume without requiring approval from the OpenLineage project.
+- Consumers can independently discover and support custom facets.
+- OpenLineage users can easily explore which producers and/or consumers best meet their compatibility needs
```
Comment (julienledem) [inline proposals/2161/registry.md:22]: SGTM

File: proposals/2161/registry.md [hunk 2]
```diff
@@ -28,6 +30,11 @@ Each consumer or producer entity can claim a name, defined in the registry: “{
 Each registered entity will provide a documentation URL for its documentation.
 The registered name is used to shorten “producer” and “schemaUrl” fields in facets.
 
+## Acceptance Guidelines
+To claim a name, an entity must have either documentation or a test/sample. "Reserving" a name prior to public functionality is discouraged.
```
Comment (julienledem) [inline proposals/2161/registry.md:34]: fair enough. +1

File: proposals/2161/registry.md [hunk 3]
```diff
@@ -28,6 +30,11 @@ Each consumer or producer entity can claim a name, defined in the registry: “{
...
+(When should one register a new facet, vs. an extension to an existing facet? Example: do I make WarningMessageRunFacet, or do I extend ErrorMessageRunFacet to have a "level" field where I can set it to "ERROR", "WARNING", "DEBUG", etc?)
```
Comment (julienledem) [inline proposals/2161/registry.md:35]: We haven't really specified how to extend an existing facet.
It's a fine line between having independent concepts and not creating something that is too verbose.
For example column lineage is separate from the schema facet and replicate some of the information in it.

File: proposals/2161/registry.md [hunk 4]
```diff
@@ -28,6 +30,11 @@ Each consumer or producer entity can claim a name, defined in the registry: “{
...
+(how does one register an extension to an existing facet?)
```
Comment (julienledem) [inline proposals/2161/registry.md:36]: We haven't defined that. Maybe just update the existing facet in a backward compatible way?

File: proposals/2161/registry.md [hunk 5]
```diff
@@ -50,13 +57,14 @@ Corresponding values to be used:
 - Generated doc from the json schemas of the facets. Publish them on openlineage.io
 
 #### Additional documentation
-We can create a documentation page per use case to document what facet can be use in what case:
+We can create a documentation page per use case to document what facet can be used in what case:
```
Comment (julienledem) [inline proposals/2161/registry.md:60]: thx!

File: proposals/2161/registry.md [hunk 6]
```diff
@@ -50,13 +57,14 @@ Corresponding values to be used:
...
 
 - Compliance with privacy laws (GDPR, CCPA, …)
 - Compliance with Banking regulation (BCBS-239)
 - Data reliability, data quality
 - Data discovery, data catalog
 - Data Governance
+- Data Lineage
```
Comment (julienledem) [inline proposals/2161/registry.md:67]: SGTM

File: proposals/2161/registry.md [hunk 7]
```diff
@@ -110,6 +118,61 @@ airflow/
...
+				"NewCustomFacet.json": {
+					"owner": "egeria",
+					"spec_versions": [ "1-0-0" ],
+					"use_cases": [ "lineage" ]
+				}
+			]
+		}
+	}
+```
+```
```
Comment (julienledem) [inline proposals/2161/registry.md:175]: Thanks for adding more examples

File: proposals/2161/registry.md [hunk 8]
```diff
@@ -210,3 +273,5 @@ Cons:
 
 - Producers need to host and maintain their own subset of the registry
 - The core repo ci cannot guarantee consistency of the registry as it relies on external references
+- The core repo cannot guarantee accuracy of registry entries (e.g. does an entity actually consume/produce the facets they say they do?)
```
Comment (julienledem) [inline proposals/2161/registry.md:276]: ok

**Review by julienledem — APPROVED** [review]

**Comment by harels** [comment]
@Sheeri would you be able to go over the responses and add comments or resolve conversations so we can merge this in?

**Comment by Sheeri** [comment]
Abaolutrly

Sent from ProtonMail mobile

-------- Original Message --------

File: proposals/2161/registry.md (hunk 3 above)
Comment (Sheeri) [reply proposals/2161/registry.md:35]: OK, will remove.

File: proposals/2161/registry.md (hunk 4 above)
Comment (Sheeri) [reply proposals/2161/registry.md:36]: I'll remove the question, and when an extension comes up we can add guidelines then :D

**Comment by Sheeri** [comment]
OK, I have resolved the comments that were effectively "thumbs-up" ones, and removed 2 open questions that we don't have answers to. cc @harels

**Comment by Sheeri** [comment]
Also resolved the merge conflicts.

**Comment by Sheeri** [comment]
(FYI the force push was because I hadn't signed off my changes with -s in my commits)

**Review by harels — APPROVED** [review]
great additions to the registry proposal, thanks for making it better!

**Pull request merged**